internal/sbi/upi/api_upi.go

```diff
@@ -43,8 +50,33 @@ func PostUpNodesLinks(c *gin.Context) {
 	for _, upf := range upi.UPFs {
 		// only associate new ones
 		if upf.UPF.UPFStatus == smf_context.NotAssociated {
+			upf.UPF.Ctx, upf.UPF.CancelFunc = context.WithCancel(context.Background())
```

Comment:
Save context and cancel function on the UPF data model so that it can be used to cancel heartbeat thread of this UPF.